builtin/logical/pki/path_tidy.go

```diff
@@ -440,6 +460,95 @@ func (b *backend) doTidyRevocationStore(ctx context.Context, req *logical.Reques
 	return nil
 }
 
+func (b *backend) doTidyExpiredIssuers(ctx context.Context, req *logical.Request, logger hclog.Logger, config *tidyConfig) error {
+	if b.System().ReplicationState().HasState(consts.ReplicationDRSecondary|consts.ReplicationPerformanceStandby) ||
+		(!b.System().LocalMount() && b.System().ReplicationState().HasState(consts.ReplicationPerformanceSecondary)) {
+		b.Logger().Debug("skipping expired issuer tidy as we're not on the primary or secondary with a local mount")
+		return nil
+	}
+
+	b.issuersLock.Lock()
+	defer b.issuersLock.Unlock()
+
+	// Fetch and parse our issuers so we have their expiration date.
+	sc := b.makeStorageContext(ctx, req.Storage)
+	issuerIDCertMap, err := fetchIssuerMapForRevocationChecking(sc)
+	if err != nil {
+		return err
+	}
+
+	// Short-circuit to avoid having to deal with the legacy mounts. While we
+	// could handle this case and remove these issuers, its somewhat
+	// unexpected behavior and we'd prefer to finish the migration first.
+	if b.useLegacyBundleCaStorage() {
+		return nil
+	}
+
+	// We want certificates which have expired before this date by a given
+	// safety buffer. So we subtract the buffer from now, and anything which
+	// has expired before our after buffer can be tidied, and anything that
+	// expired after this buffer must be kept.
+	now := time.Now()
+	afterBuffer := now.Add(-1 * config.IssuerSafetyBuffer)
+
+	rebuildChainsAndCRL := false
+
+	for issuer, cert := range issuerIDCertMap {
+		if cert.NotAfter.After(afterBuffer) {
+			continue
+		}
+
+		entry, err := sc.fetchIssuerById(issuer)
+		if err != nil {
+			return nil
+		}
+
+		// This issuer's certificate has expired. We explicitly persist the
+		// key, but log both the certificate and the keyId to the
+		// informational logs so an admin can recover the removed cert if
+		// necessary or remove the key (and know which cert it belonged to),
+		// if desired.
+		msg := "[Tidy on mount: %v] Issuer %v has expired by %v and is being removed.\n"
+		msg += "Serial Number: %v\n"
+		msg += "Key ID (if any): %v\n"
+		msg += "Certificate:\n%v\n"
+		idAndName := fmt.Sprintf("[id:%v/name:%v]", entry.ID, entry.Name)
+		msg = fmt.Sprintf(msg, b.backendUUID, idAndName, config.IssuerSafetyBuffer, entry.SerialNumber, entry.KeyID, entry.Certificate)
+
+		b.Logger().Info(msg)
+
+		wasDefault, err := sc.deleteIssuer(issuer)
+		if err != nil {
+			b.Logger().Error(fmt.Sprintf("failed to remove %v: %v", idAndName, err))
+			return err
+		}
+		if wasDefault {
+			b.Logger().Warn(fmt.Sprintf("expired issuer %v was default; it is strongly encouraged to choose a new default issuer for backwards compatibility", idAndName))
```

Comment:
I'm worried that someone will shoot themselves in the foot with this behaviour, including with the existing (consul) use case.  I think it would be kind to error-return the API call to tidy issuers when the default would be tidied, and allow flags "delete default issuer" and "migrate default from expired issuer" to tidy.  I'm really worried that logging just wouldn't be seen by anyone until after everything stops working.

Reply:
Hmmmm, that does rather complicate the API. To fix the Consul thing, they'll handle updating the default issuer, so I don't think this'll be an issuer there (once fixed). Plus, with the 1y post-expiry default, I don't think this'll still be a problem that much later.

Reply:
This also won't be backported, we'll just target the latest version for this.